File: sdks/local-rp/java/src/test/java/community/catalyst/linkkeys/localrp/IdentityTest.java
package community.catalyst.linkkeys.localrp;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.time.Duration;
import java.time.Instant;
import java.util.List;

import org.junit.jupiter.api.Test;

import community.catalyst.linkkeys.localrp.crypto.AeadSuite;
import community.catalyst.linkkeys.localrp.crypto.Crypto;
import community.catalyst.linkkeys.localrp.wire.Codec;
import community.catalyst.linkkeys.localrp.wire.Types.LocalRpDescriptor;

/** Unit tests for {@link Identity} (mirrors the Rust/Go SDKs' own {@code identity} module tests). */
class IdentityTest {

    private static Identity.LocalRpKeyMaterial material() {
        return Identity.generateLocalRpIdentity(new Identity.GenerateLocalRpIdentityConfig("Test App", Instant.now()));
    }

    @Test
    void generateIdentityDefaultsBothSuitesAndTenYearLifetime() {
        Identity.LocalRpKeyMaterial identity = material();
        assertEquals(64, identity.fingerprint().length());
        assertEquals(Crypto.fingerprint(identity.signingPublicKey()), identity.fingerprint());

        LocalRpDescriptor descriptor = Codec.decodeLocalRpDescriptor(identity.descriptor().descriptor());
        assertEquals("Test App", descriptor.appName());
        assertEquals(AeadSuite.allSupported(), descriptor.supportedSuites());
    }

    @Test
    void generateIdentityRejectsEmptyAppName() {
        assertThrows(
                SdkException.class,
                () -> Identity.generateLocalRpIdentity(new Identity.GenerateLocalRpIdentityConfig("", Instant.now())));
    }

    @Test
    void generateIdentityRejectsEmptySuiteList() {
        Identity.GenerateLocalRpIdentityConfig config =
                new Identity.GenerateLocalRpIdentityConfig("Test App", Instant.now());
        config.supportedSuites = List.of();
        assertThrows(SdkException.class, () -> Identity.generateLocalRpIdentity(config));
    }

    @Test
    void signingAndEncryptionKeyByteRoundTrips() {
        byte[] key = new byte[32];
        java.util.Arrays.fill(key, (byte) 7);
        assertArrayEquals(key, Identity.signingKeyFromBytes(Identity.signingKeyToBytes(key)));
        assertArrayEquals(key, Identity.encryptionKeyFromBytes(Identity.encryptionKeyToBytes(key)));
        assertThrows(SdkException.class, () -> Identity.signingKeyFromBytes(new byte[31]));
        assertThrows(SdkException.class, () -> Identity.encryptionKeyFromBytes(new byte[33]));
    }

    @Test
    void fingerprintStringRoundTripValidatesHex() {
        Identity.LocalRpKeyMaterial identity = material();
        String s = Identity.fingerprintToString(identity.fingerprint());
        assertEquals(identity.fingerprint(), Identity.fingerprintFromString(s));
        assertThrows(SdkException.class, () -> Identity.fingerprintFromString("not-hex"));
        assertThrows(SdkException.class, () -> Identity.fingerprintFromString("a".repeat(63)));
    }

    @Test
    void identityBundleByteRoundTrip() {
        Identity.LocalRpKeyMaterial identity = material();
        byte[] bytes = Identity.localRpIdentityToBytes(identity);
        Identity.LocalRpKeyMaterial roundTripped = Identity.localRpIdentityFromBytes(bytes);

        assertArrayEquals(identity.signingPrivateKey(), roundTripped.signingPrivateKey());
        assertArrayEquals(identity.signingPublicKey(), roundTripped.signingPublicKey());
        assertArrayEquals(identity.encryptionPrivateKey(), roundTripped.encryptionPrivateKey());
        assertArrayEquals(identity.encryptionPublicKey(), roundTripped.encryptionPublicKey());
        assertEquals(identity.fingerprint(), roundTripped.fingerprint());
        assertArrayEquals(identity.descriptor().descriptor(), roundTripped.descriptor().descriptor());
        assertArrayEquals(identity.descriptor().signature(), roundTripped.descriptor().signature());
    }

    @Test
    void identityBundleRejectsBadMagicAndTruncation() {
        Identity.LocalRpKeyMaterial identity = material();
        byte[] bytes = Identity.localRpIdentityToBytes(identity);
        byte[] badMagic = bytes.clone();
        badMagic[0] ^= (byte) 0xff;
        assertThrows(SdkException.class, () -> Identity.localRpIdentityFromBytes(badMagic));

        byte[] truncated = java.util.Arrays.copyOfRange(bytes, 0, 10);
        assertThrows(SdkException.class, () -> Identity.localRpIdentityFromBytes(truncated));
    }

    @Test
    void checkExpirationsWrapsThresholds() {
        Identity.GenerateLocalRpIdentityConfig config =
                new Identity.GenerateLocalRpIdentityConfig("Test App", Instant.now());
        config.lifetime = Duration.ofDays(100);
        Identity.LocalRpKeyMaterial identity = Identity.generateLocalRpIdentity(config);

        LocalRp.ExpirationStatus status = LinkKeysLocalRp.checkExpirations(identity, Instant.now());
        assertEquals(LocalRp.ExpirationLevel.NOTICE, status.level());

        Instant farFuture = Instant.now().plus(Duration.ofDays(200));
        LocalRp.ExpirationStatus expired = LinkKeysLocalRp.checkExpirations(identity, farFuture);
        assertEquals(LocalRp.ExpirationLevel.EXPIRED, expired.level());
    }
}

File: sdks/local-rp/kotlin/src/test/kotlin/community/catalyst/linkkeys/localrp/kt/IdentityTest.kt
package community.catalyst.linkkeys.localrp.kt

import java.time.Duration
import java.time.Instant
import org.junit.jupiter.api.Assertions.assertArrayEquals
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Test
import community.catalyst.linkkeys.localrp.kt.protocol.Crypto

/** Unit tests for [generateLocalRpIdentity] and the byte-storage helpers (mirrors the Java/Rust/Go SDKs' own identity module tests). */
class IdentityTest {

    private fun material(): LocalRpIdentity = generateLocalRpIdentity(appName = "Test App", now = Instant.now())

    @Test
    fun generateIdentityDefaultsBothSuitesAndTenYearLifetime() {
        val identity = material()
        assertEquals(64, identity.fingerprint.length)
        assertEquals(Crypto.fingerprint(identity.signingPublicKey), identity.fingerprint)
        assertEquals(Duration.ofDays(3650), LocalRpIdentity.DEFAULT_LIFETIME)
    }

    @Test
    fun generateIdentityRejectsEmptyAppName() {
        assertThrows(LocalRpException.InvalidInput::class.java) { generateLocalRpIdentity(appName = "", now = Instant.now()) }
    }

    @Test
    fun generateIdentityRejectsEmptySuiteList() {
        assertThrows(LocalRpException.InvalidInput::class.java) {
            generateLocalRpIdentity(appName = "Test App", now = Instant.now(), supportedSuites = emptyList())
        }
    }

    @Test
    fun signingAndEncryptionKeyByteRoundTrips() {
        val key = ByteArray(32) { 7 }
        assertArrayEquals(key, signingKeyFromBytes(signingKeyToBytes(key)))
        assertArrayEquals(key, encryptionKeyFromBytes(encryptionKeyToBytes(key)))
        assertThrows(LocalRpException.InvalidInput::class.java) { signingKeyFromBytes(ByteArray(31)) }
        assertThrows(LocalRpException.InvalidInput::class.java) { encryptionKeyFromBytes(ByteArray(33)) }
    }

    @Test
    fun fingerprintStringRoundTripValidatesHex() {
        val identity = material()
        val s = fingerprintToString(identity.fingerprint)
        assertEquals(identity.fingerprint, fingerprintFromString(s))
        assertThrows(LocalRpException.InvalidInput::class.java) { fingerprintFromString("not-hex") }
        assertThrows(LocalRpException.InvalidInput::class.java) { fingerprintFromString("a".repeat(63)) }
    }

    @Test
    fun identityBundleByteRoundTrip() {
        val identity = material()
        val bytes = identity.toBytes()
        val roundTripped = localRpIdentityFromBytes(bytes)

        assertArrayEquals(identity.signingPrivateKey, roundTripped.signingPrivateKey)
        assertArrayEquals(identity.signingPublicKey, roundTripped.signingPublicKey)
        assertArrayEquals(identity.encryptionPrivateKey, roundTripped.encryptionPrivateKey)
        assertArrayEquals(identity.encryptionPublicKey, roundTripped.encryptionPublicKey)
        assertEquals(identity.fingerprint, roundTripped.fingerprint)
        assertEquals(identity, roundTripped)
    }

    @Test
    fun identityBundleRejectsBadMagicAndTruncation() {
        val identity = material()
        val bytes = identity.toBytes()
        val badMagic = bytes.clone()
        badMagic[0] = (badMagic[0].toInt() xor 0xff).toByte()
        assertThrows(LocalRpException.InvalidInput::class.java) { localRpIdentityFromBytes(badMagic) }

        val truncated = bytes.copyOfRange(0, 10)
        assertThrows(LocalRpException.InvalidInput::class.java) { localRpIdentityFromBytes(truncated) }
    }

    @Test
    fun checkExpirationsWrapsThresholds() {
        val identity = generateLocalRpIdentity(appName = "Test App", now = Instant.now(), lifetime = Duration.ofDays(100))

        val status = checkExpirations(identity, Instant.now())
        assertEquals(ExpirationLevel.NOTICE, status.level)

        val farFuture = Instant.now().plus(Duration.ofDays(200))
        val expired = checkExpirations(identity, farFuture)
        assertEquals(ExpirationLevel.EXPIRED, expired.level)
    }
}
